043-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,112 @@
+# Liveness and Readiness probes in a KRaft Kafka cluster
+
+This proposal describes the liveness and readiness probes that should be put in place 
+for a Kafka cluster that is using KRaft rather than ZooKeeper. This includes a KRaft 
+cluster with combined nodes and one that has separate controller and broker nodes.
+
+## Current situation
+
+These are the current probes used in Strimzi when ZooKeeper is present:
+
+|Component| Liveness | Readiness |
+|---|---|---|
+| ZooKeeper | "imok" from 127.0.0.1:12181 | "imok" from 127.0.0.1:12181 |
+| ZK-based broker | if (ready) { listening on 9091 } else { have ZK session } | if (ready) { return true } |
+
+The "ready" check in both cases is based on the existence of the file `../kafka-ready` which is 
+created by the KafkaAgent when the broker state metric has value 3 (RUNNING).
+
+## Background
+
+This section contains some background information about the ports and metrics we could use for checks.
+
+### Ports
+In KRaft mode there are two different ports that the node could be listening on.
+If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is currently hard-coded to 9090.
+If it has a broker process it should be listening on the replication port which is hard-coded to 9091.
+If the node is running as a combined mode, it should be listening on both ports
+
+### BrokerState
+The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions are:
+
+NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)
+
+The other possible value is 127 which represents UNKNOWN.
+
+### Current state metric and quorum
+
+Some useful things to be aware of with the current-state metric and how the controller quorum affects the broker nodes:
+
+* The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates whether a controller is currently a leader or follower.
+If this metric shows one of these two states, this means the leader election has happened and a controller quorum has been successfully formed.
+* The current-state metric is not accessible to the existing KafkaAgent because it is not a Yammer metric.
+* The broker nodes will not move to a RUNNING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a combined node, 
+then for a broker and then for a controller.
+
+### Combined mode
+
+The proposed probes are:
+
+* Liveness: node is listening on 9090 and 9091
+* Readiness: node is listening on 9090 and 9091
```

Comment:
Adding comment from @scholzj that got lost in the file rename:

[scholzj]
I really wonder if there is a situation where 9090 will fail and 9091 will pass for example. Do we really need to check both?

[katheris]
The ports are used for two different things, so I think for a combined mode to work correctly it would need to be listening on both. That being said, by looking at the code and experimenting I haven't found a scenario where it would get to the point of listening on the replication port (9091) and not be listening on 9090, since the controller starts up first and then starts the broker. So theoretically (as far as I can tell) we could only check 9091, but is it much overhead to just check both?

Reply:
Having looked closely at the `BrokerServer.scala` code (which is used for the `KafkaRaftServer`, i.e. KRaft mode, not ZK model) in `startup()` it:

1. creates the LogManager, but does not start it (i.e. log recovery is deferred)
2. instantiates the `SocketServer` (which results in the `ServerSocket#bind()` call, which results in the `bind` and `listen` syscalls, which result in the kernel putting the socket in the `LISTEN` state).
3. Much later in `startup()` it instantiates the `BrokerMetadataPublisher` which calls `LogManager#startup()` (initiating log recovery) as a side effect of handling Raft commit or snapshots 
4. Only then does `socketServer.enableRequestProcessing()` get called and the broker starts actually accepting queued client requests.

In other words, for a KRaft broker $\texttt{LISTEN} \mathrel{\rlap{\hskip .5em/}}\Longrightarrow brokerState > RECOVERY$. Which is different from a ZK-based broker. The question is how much this is really a problem, given a KRaft server with `process.roles=broker` explicitly binds and listens before starting recovery. So long as the broker can serve `Metadata` requests (which is will be able to do by step 4, I think) the rest of the Kafka protocol should guarantee that a client should be able to bootstrap. It will (eventually) find that the broker isn't the leader for any partitions (while that broker is in log recovery). 

In other words, I'm thinking that perhaps the readiness check we should be using for KRaft (controller, broker or combined) is not on the broker state or server socket LISTEN state, but rather whether we can make a TCP connection from a client (i.e. a client socket gets into the `ESTABLISHED` TCP socket state).

Reply:
If I understand what you've said above and the code correctly, for a server with `process.roles=broker` there are three events that happen that we could check readiness on. First it starts listening on the replication port, then (after loading metadata from the controller quorum) it starts accepting TCP connections, then finally it finishes log recovery and moves to a RUNNING state. And it's only at the final step that the broker is actually eligible to be a leader for any topic-partitions. Is that correct?

If so, wouldn't checking for TCP connections mean we mark the broker as ready when it is still doing log recovery? I guess this is at least a better check than just "is it listening", but I still feel unsure about what the risks are of marking a broker as ready if it's still in log recovery. I guess it would mean we wouldn't have the problem where the roller tries to roll the brokers while they are doing recovery?

For the combined mode, from my understanding of the above, if we mark the server as ready only when it starts accepting TCP connections to the replication port, then we would still hit the problem where if all the pods had been pending then a single server by itself can't become ready because there is no controller quorum so the broker endpoint would never start accepting TCP connections right?

Reply:
> For combined mode...

This is an excellent point. I think it's true that a controller will connect itself to participate in Raft, which it can only do if it's ready. In the single node case this will create a cyclic dependency. But it's probably also a problem even when #nodes > 1 (e.g. it might prevent the election of a leader if another voter is down and we a vote for ourselves would win us leadership but we can't connect). Perhaps we should simply not tie readiness to network.